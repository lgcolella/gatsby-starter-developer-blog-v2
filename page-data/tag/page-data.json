{"componentChunkName":"component---src-pages-tag-index-tsx","path":"/tag/","webpackCompilationHash":"3c63780dbbd702bc3d73","result":{"data":{"allMarkdownRemark":{"edges":[{"node":{"frontmatter":{"tags":["javascript","react","electron","nodejs"]}}},{"node":{"frontmatter":{"tags":["javascript","angular","sass"]}}},{"node":{"frontmatter":{"tags":["laravel","vuejs","typescript","javascript","nodejs"]}}},{"node":{"frontmatter":{"tags":["typescript","javascript","rxjs"]}}}]},"allFile":{"edges":[{"node":{"name":"angular","childImageSharp":{"fluid":{"aspectRatio":0.9375,"src":"/static/62506c6b62b9db1eb2e8fef5c70bcf62/b60ef/angular.png","srcSet":"/static/62506c6b62b9db1eb2e8fef5c70bcf62/b60ef/angular.png 90w","sizes":"(max-width: 90px) 100vw, 90px"}}}},{"node":{"name":"javascript","childImageSharp":{"fluid":{"aspectRatio":1,"src":"/static/7f130501681cb3d0129f266820c78b24/9d33d/javascript.png","srcSet":"/static/7f130501681cb3d0129f266820c78b24/ca492/javascript.png 100w,\n/static/7f130501681cb3d0129f266820c78b24/92ebe/javascript.png 200w,\n/static/7f130501681cb3d0129f266820c78b24/9d33d/javascript.png 400w,\n/static/7f130501681cb3d0129f266820c78b24/bd9e5/javascript.png 600w,\n/static/7f130501681cb3d0129f266820c78b24/a056a/javascript.png 800w,\n/static/7f130501681cb3d0129f266820c78b24/31688/javascript.png 1052w","sizes":"(max-width: 400px) 100vw, 400px"}}}},{"node":{"name":"laravel","childImageSharp":{"fluid":{"aspectRatio":1,"src":"/static/b0f8da922e4a3d2ea26301f115ffc5a7/9d33d/laravel.png","srcSet":"/static/b0f8da922e4a3d2ea26301f115ffc5a7/ca492/laravel.png 100w,\n/static/b0f8da922e4a3d2ea26301f115ffc5a7/92ebe/laravel.png 200w,\n/static/b0f8da922e4a3d2ea26301f115ffc5a7/9d33d/laravel.png 400w,\n/static/b0f8da922e4a3d2ea26301f115ffc5a7/bd9e5/laravel.png 600w,\n/static/b0f8da922e4a3d2ea26301f115ffc5a7/a056a/laravel.png 800w,\n/static/b0f8da922e4a3d2ea26301f115ffc5a7/55230/laravel.png 1200w","sizes":"(max-width: 400px) 100vw, 400px"}}}},{"node":{"name":"rxjs","childImageSharp":{"fluid":{"aspectRatio":1,"src":"/static/cf61f21dd1e03663dcfc29d520894456/9d33d/rxjs.png","srcSet":"/static/cf61f21dd1e03663dcfc29d520894456/ca492/rxjs.png 100w,\n/static/cf61f21dd1e03663dcfc29d520894456/92ebe/rxjs.png 200w,\n/static/cf61f21dd1e03663dcfc29d520894456/9d33d/rxjs.png 400w,\n/static/cf61f21dd1e03663dcfc29d520894456/601ae/rxjs.png 580w","sizes":"(max-width: 400px) 100vw, 400px"}}}},{"node":{"name":"typescript","childImageSharp":{"fluid":{"aspectRatio":1,"src":"/static/26cc95f255ccb936d154b43614f61602/9d33d/typescript.png","srcSet":"/static/26cc95f255ccb936d154b43614f61602/ca492/typescript.png 100w,\n/static/26cc95f255ccb936d154b43614f61602/92ebe/typescript.png 200w,\n/static/26cc95f255ccb936d154b43614f61602/9d33d/typescript.png 400w,\n/static/26cc95f255ccb936d154b43614f61602/bd9e5/typescript.png 600w,\n/static/26cc95f255ccb936d154b43614f61602/a056a/typescript.png 800w,\n/static/26cc95f255ccb936d154b43614f61602/934ff/typescript.png 1187w","sizes":"(max-width: 400px) 100vw, 400px"}}}},{"node":{"name":"nodejs","childImageSharp":{"fluid":{"aspectRatio":1,"src":"/static/6dcf699938b2291d1634f91adb0dded1/9d33d/nodejs.png","srcSet":"/static/6dcf699938b2291d1634f91adb0dded1/ca492/nodejs.png 100w,\n/static/6dcf699938b2291d1634f91adb0dded1/92ebe/nodejs.png 200w,\n/static/6dcf699938b2291d1634f91adb0dded1/9d33d/nodejs.png 400w,\n/static/6dcf699938b2291d1634f91adb0dded1/bd9e5/nodejs.png 600w,\n/static/6dcf699938b2291d1634f91adb0dded1/a056a/nodejs.png 800w","sizes":"(max-width: 400px) 100vw, 400px"}}}},{"node":{"name":"sass","childImageSharp":{"fluid":{"aspectRatio":1.3333333333333333,"src":"/static/eac0fdba26c36d995b451da52e901668/9d33d/sass.png","srcSet":"/static/eac0fdba26c36d995b451da52e901668/ca492/sass.png 100w,\n/static/eac0fdba26c36d995b451da52e901668/92ebe/sass.png 200w,\n/static/eac0fdba26c36d995b451da52e901668/9d33d/sass.png 400w,\n/static/eac0fdba26c36d995b451da52e901668/bd9e5/sass.png 600w,\n/static/eac0fdba26c36d995b451da52e901668/a056a/sass.png 800w,\n/static/eac0fdba26c36d995b451da52e901668/4f941/sass.png 1280w","sizes":"(max-width: 400px) 100vw, 400px"}}}},{"node":{"name":"electron","childImageSharp":{"fluid":{"aspectRatio":0.922883487007544,"src":"/static/25a28f18a44f6c974e80a257dda94881/9d33d/electron.png","srcSet":"/static/25a28f18a44f6c974e80a257dda94881/ca492/electron.png 100w,\n/static/25a28f18a44f6c974e80a257dda94881/92ebe/electron.png 200w,\n/static/25a28f18a44f6c974e80a257dda94881/9d33d/electron.png 400w,\n/static/25a28f18a44f6c974e80a257dda94881/bd9e5/electron.png 600w,\n/static/25a28f18a44f6c974e80a257dda94881/a056a/electron.png 800w,\n/static/25a28f18a44f6c974e80a257dda94881/5d8d3/electron.png 2202w","sizes":"(max-width: 400px) 100vw, 400px"}}}},{"node":{"name":"react","childImageSharp":{"fluid":{"aspectRatio":1,"src":"/static/f87e161375f53a8acb0caa77ea0ef48e/9d33d/react.png","srcSet":"/static/f87e161375f53a8acb0caa77ea0ef48e/ca492/react.png 100w,\n/static/f87e161375f53a8acb0caa77ea0ef48e/92ebe/react.png 200w,\n/static/f87e161375f53a8acb0caa77ea0ef48e/9d33d/react.png 400w,\n/static/f87e161375f53a8acb0caa77ea0ef48e/bd9e5/react.png 600w,\n/static/f87e161375f53a8acb0caa77ea0ef48e/a056a/react.png 800w","sizes":"(max-width: 400px) 100vw, 400px"}}}},{"node":{"name":"vuejs","childImageSharp":{"fluid":{"aspectRatio":1.158511269276394,"src":"/static/bf9b350fe064bef380dfc2a3cff449f8/9d33d/vuejs.png","srcSet":"/static/bf9b350fe064bef380dfc2a3cff449f8/ca492/vuejs.png 100w,\n/static/bf9b350fe064bef380dfc2a3cff449f8/92ebe/vuejs.png 200w,\n/static/bf9b350fe064bef380dfc2a3cff449f8/9d33d/vuejs.png 400w,\n/static/bf9b350fe064bef380dfc2a3cff449f8/bd9e5/vuejs.png 600w,\n/static/bf9b350fe064bef380dfc2a3cff449f8/a056a/vuejs.png 800w,\n/static/bf9b350fe064bef380dfc2a3cff449f8/dbefb/vuejs.png 7813w","sizes":"(max-width: 400px) 100vw, 400px"}}}}]}},"pageContext":{"isCreatedByStatefulCreatePages":true,"slug":"/tag/","langKey":"en"}}}