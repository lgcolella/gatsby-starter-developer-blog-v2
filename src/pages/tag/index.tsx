/* Vendor imports */
import React from 'react'
import PropTypes from 'prop-types'
import { graphql, Link } from 'gatsby'
import Image from 'gatsby-image'
/* App imports */
import { ChildImageSharpFluid, AvailableTag } from '../../types'
import Layout from '../../components/layout'
import SEO from '../../components/seo'
import Config from '../../../config'
import Utils from '../../utils'
import Lang from '../../i18n/default'
const style = require('./tag.module.less')

const Tag = ({ data }: Props) => {
  const rawTags = data.allMarkdownRemark.edges
    .map(edge => edge.node.frontmatter.tags)
    .reduce((prev, curr) => prev.concat(curr))
  const tags = rawTags
    .filter((tag, index) => index === rawTags.indexOf(tag))
    .sort() // Remove duplicates and sort values
  const tagPage = Config.pages.tag
  const tagImages = data.allFile.edges.reduce(
    (prevResult, currEdge) => {
      return Object.assign({}, prevResult, {
        [currEdge.node.name]: currEdge.node.childImageSharp.fluid,
      })
    },
    {} as any
  )

  return (
    <Layout title={Lang.pages.tag.title}>
      <SEO
        title={Lang.pages.tag.title}
        description={Lang.pages.tag.description}
        path={tagPage}
      />
      <div>
        {tags.map(tag => (
          <Link
            to={Utils.resolveUrl(tagPage, tag)}
            className={style.card}
            key={tag}
          >
            <div className={style.cover}>
              <Image fluid={tagImages[tag]} loading={'eager'} />
            </div>
            <div>
              <h3>{Lang.tags[tag].name || Utils.capitalize(tag)}</h3>
              <p>{Lang.tags[tag].description}</p>
              <label>
                {rawTags.filter(sTag => sTag === tag).length +
                  ' ' +
                  Lang.pages.tag.postsLabel}
              </label>
            </div>
          </Link>
        ))}
      </div>
    </Layout>
  )
}

Tag.propTypes = {
  data: PropTypes.shape({
    allMarkdownRemark: PropTypes.shape({
      edges: PropTypes.arrayOf(
        PropTypes.shape({
          node: PropTypes.shape({
            frontmatter: PropTypes.shape({
              tags: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
            }).isRequired,
          }).isRequired,
        }).isRequired
      ).isRequired,
    }).isRequired,
    allFile: PropTypes.shape({
      edges: PropTypes.arrayOf(
        PropTypes.shape({
          node: PropTypes.shape({
            name: PropTypes.string.isRequired,
            childImageSharp: PropTypes.shape({
              fluid: PropTypes.object.isRequired,
            }).isRequired,
          }).isRequired,
        }).isRequired
      ).isRequired,
    }).isRequired,
  }).isRequired,
}

interface Props {
  data: {
    allMarkdownRemark: {
      edges: Array<{
        node: {
          frontmatter: {
            tags: AvailableTag[]
          }
        }
      }>
    }
    allFile: {
      edges: Array<{
        node: {
          name: AvailableTag
          childImageSharp: ChildImageSharpFluid
        }
      }>
    }
  }
}

export const query = graphql`
  {
    allMarkdownRemark(filter: { fileAbsolutePath: { regex: "/index.md$/" } }) {
      edges {
        node {
          frontmatter {
            tags
          }
        }
      }
    }
    allFile(filter: { relativeDirectory: { eq: "tags" } }) {
      edges {
        node {
          name
          childImageSharp {
            fluid(maxWidth: 400) {
              ...GatsbyImageSharpFluid_noBase64
            }
          }
        }
      }
    }
  }
`

export default Tag
