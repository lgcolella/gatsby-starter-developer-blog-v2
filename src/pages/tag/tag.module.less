@import (reference) '../../style/index.less';

.card {
    .flex-center();
    border-bottom: 2px solid @theme-title;
    box-sizing: border-box;
    flex-direction: row-reverse;
    flex-wrap: nowrap;
    height: auto;
    justify-content: left;
    margin-bottom: @margin-medium;
    padding: @padding-medium @padding-zero @padding-medium * 2 @padding-zero;
    width: 100%;
    h3 {
        margin: @margin-zero;
        text-align: left;
    }
    p {
        text-align: left;
        padding-right: @padding-medium;
    }
    :hover {
        h3, p {
            text-decoration: underline;
        }
    }
    @media @mobile {
        flex-direction: column;
        margin: @margin-zero;
        padding: @padding-medium @padding-zero;
        p {
            padding-right: @padding-zero;
        }
    }
}

.cover {
    @width: 10%;

    min-width: @width;
    margin-left: auto;
    @media @mobile {
        margin-right: auto;
        margin-bottom: @margin-medium;
        margin-left: @margin-zero;
        min-width: @width * 2.5;
    }
}
