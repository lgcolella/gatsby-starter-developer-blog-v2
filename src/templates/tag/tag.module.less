@import (reference) '../../style/index.less';

.heading {
    align-items: center;
    display: flex;
    justify-content: space-between;
    h1 {
        margin: @margin-zero;
    }
}

.cover {
    float: left;
    margin-left: @margin-medium;
    width: 8%;
    @media @mobile {
        width: 25%;
    }
}