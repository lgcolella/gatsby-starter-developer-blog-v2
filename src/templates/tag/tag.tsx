/* Vendor imports */
import React from 'react'
import PropTypes from 'prop-types'
import { graphql } from 'gatsby'
import Img from 'gatsby-image'
/* App imports */
import { ChildImageSharpFluid, AvailableTag } from '../../types'
import Layout from '../../components/layout'
import SEO from '../../components/seo'
import PostList from '../../components/post-list'
import Config from '../../../config'
import Utils from '../../utils'
import Lang from '../../i18n/default'
const style = require('./tag.module.less')

const TagPage = ({ data, pageContext }: Props) => {
  const tag = pageContext.tag
  const tagName = Lang.tags[tag].name || Utils.capitalize(tag)
  const tagImage = (() => {
    let edge = data.allFile.edges.find(edge => edge.node.name === tag)
    if (edge) {
      return edge.node.childImageSharp.fluid
    } else {
      throw 'Not image file found for tag: ' + tag
    }
  })()

  return (
    <Layout>
      <SEO
        title={tagName}
        description={`${Lang.templates.tag.pageTitle} ${tagName}`}
        path={Utils.resolveUrl(Config.pages.tag, tag)}
        keywords={[tagName]}
      />
      <div className={style.heading}>
        <div>
          <h1>{tagName}</h1>
        </div>
        <div className={style.cover}>
          <Img fluid={tagImage} loading={'eager'} />
        </div>
      </div>
      <PostList posts={data.allMarkdownRemark.edges} />
    </Layout>
  )
}

TagPage.propTypes = {
  data: PropTypes.shape({
    allMarkdownRemark: PropTypes.shape({
      edges: PropTypes.arrayOf(PropTypes.object).isRequired,
    }).isRequired,
    allFile: PropTypes.shape({
      edges: PropTypes.arrayOf(
        PropTypes.shape({
          node: PropTypes.shape({
            name: PropTypes.string.isRequired,
            childImageSharp: PropTypes.shape({
              fluid: PropTypes.object.isRequired,
            }).isRequired,
          }).isRequired,
        })
      ).isRequired,
    }).isRequired,
  }).isRequired,
  pageContext: PropTypes.shape({
    tag: PropTypes.string.isRequired,
  }).isRequired,
}

export interface Props {
  data: {
    allMarkdownRemark: {
      edges: Array<{
        node: {
          frontmatter: {
            title: string
            date: string
            path: string
            tags: AvailableTag[]
            excerpt: string
            cover: {
              childImageSharp: ChildImageSharpFluid
            }
          }
        }
      }>
    }
    allFile: {
      edges: Array<{
        node: {
          name: AvailableTag
          childImageSharp: ChildImageSharpFluid
        }
      }>
    }
  }
  pageContext: {
    tag: AvailableTag
  }
}

export const pageQuery = graphql`
  query($tag: String!) {
    allMarkdownRemark(
      filter: {
        frontmatter: { tags: { in: [$tag] } }
        fileAbsolutePath: { regex: "/index.md$/" }
      }
      sort: { fields: [frontmatter___date], order: DESC }
    ) {
      edges {
        node {
          frontmatter {
            title
            date(formatString: "YYYY-MM-DD")
            path
            tags
            excerpt
            cover {
              childImageSharp {
                fluid(maxWidth: 600) {
                  ...GatsbyImageSharpFluid_tracedSVG
                }
              }
            }
          }
        }
      }
    }
    allFile(filter: { name: { eq: $tag }, dir: { regex: "/tags$/" } }) {
      edges {
        node {
          name
          childImageSharp {
            fluid(maxHeight: 200) {
              ...GatsbyImageSharpFluid_noBase64
            }
          }
        }
      }
    }
  }
`

export default TagPage
