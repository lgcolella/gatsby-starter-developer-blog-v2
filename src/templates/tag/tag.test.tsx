/* Vendor imports */
import React from 'react'
import { shallow } from 'enzyme'
const Chance = require('chance')()
/* App imports */
import TagPage from './index'
import { Props } from './tag'
import PostList from '../../components/post-list'
import Lang from '../../i18n/default'
import { ImageSharpFluid } from '../../../tests/mocks'

describe('Tag template', () => {
  const TagProps: Props = {
    data: {
      allMarkdownRemark: {
        edges: [
          // Article #1
          {
            node: {
              frontmatter: {
                title: 'Title of article #1',
                path: 'path-to-the-article-1',
                tags: ['javascript', 'express'],
                date: '2019-05-14',
                excerpt: 'an excertp for the article #1',
                cover: {
                  childImageSharp: ImageSharpFluid,
                },
              },
            },
          },
          // Article #2
          {
            node: {
              frontmatter: {
                title: 'Title of article #2',
                path: 'path-to-the-article-2',
                tags: ['nodejs', 'angular'],
                date: '2018-12-15',
                excerpt: 'an excertp for the article #2',
                cover: {
                  childImageSharp: ImageSharpFluid,
                },
              },
            },
          },
        ],
      },
      allFile: {
        edges: [
          {
            node: {
              name: 'javascript',
              childImageSharp: ImageSharpFluid,
            },
          },
          {
            node: {
              name: 'angular',
              childImageSharp: ImageSharpFluid,
            },
          },
        ],
      },
    },
    pageContext: {
      tag: 'angular',
    },
  }

  test('matches snapshot', () => {
    let wrapper = shallow(<TagPage {...TagProps} />)

    expect(wrapper).toMatchSnapshot()
  })

  test('shows only posts with the passed tag', () => {
    let tagName = Chance.string()
    let componentProps = Object.assign({}, TagProps)

    for (let i = 0; i < 5; i++) {
      componentProps.data.allMarkdownRemark.edges.push({
        node: {
          frontmatter: {
            title: Chance.string(),
            path: Chance.url(),
            tags: Chance.pickset(Object.keys(Lang.tags).push(tagName)),
            date: Chance.date({ string: true }),
            excerpt: Chance.sentence(),
            cover: {
              childImageSharp: ImageSharpFluid,
            },
          },
        },
      })
    }

    let wrapper = shallow(<TagPage {...componentProps} />)
    let posts = wrapper.find(PostList).prop('posts')

    expect(componentProps.data.allMarkdownRemark.edges).toEqual(posts)
  })
})
