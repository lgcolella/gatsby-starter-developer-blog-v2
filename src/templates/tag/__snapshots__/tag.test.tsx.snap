// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag template matches snapshot 1`] = `
<Layout
  className=""
  showHeader={true}
  title=""
>
  <SEO
    description="All posts about Angular"
    keywords={
      Array [
        "Angular",
      ]
    }
    path="/tag/angular/"
    title="Angular"
  />
  <div
    className="heading"
  >
    <div>
      <h1>
        Angular
      </h1>
    </div>
    <div
      className="cover"
    >
      <Image
        Tag="div"
        alt=""
        durationFadeIn={500}
        fadeIn={true}
        fluid={
          Object {
            "aspectRatio": 1.7,
            "sizes": "(max-width: 600px) 100vw, 600px",
            "src": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
            "srcSet": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
          }
        }
        loading="eager"
      />
    </div>
  </div>
  <PostList
    contentClassName=""
    posts={
      Array [
        Object {
          "node": Object {
            "frontmatter": Object {
              "cover": Object {
                "childImageSharp": Object {
                  "fluid": Object {
                    "aspectRatio": 1.7,
                    "sizes": "(max-width: 600px) 100vw, 600px",
                    "src": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
                    "srcSet": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
                  },
                },
              },
              "date": "2019-05-14",
              "excerpt": "an excertp for the article #1",
              "path": "path-to-the-article-1",
              "tags": Array [
                "javascript",
                "express",
              ],
              "title": "Title of article #1",
            },
          },
        },
        Object {
          "node": Object {
            "frontmatter": Object {
              "cover": Object {
                "childImageSharp": Object {
                  "fluid": Object {
                    "aspectRatio": 1.7,
                    "sizes": "(max-width: 600px) 100vw, 600px",
                    "src": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
                    "srcSet": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
                  },
                },
              },
              "date": "2018-12-15",
              "excerpt": "an excertp for the article #2",
              "path": "path-to-the-article-2",
              "tags": Array [
                "nodejs",
                "angular",
              ],
              "title": "Title of article #2",
            },
          },
        },
      ]
    }
  />
</Layout>
`;
